org.eclipse.scanning.test/src/org/eclipse/scanning/test/scan/nexus/PosDetectorScanTest.java

```diff
@@ -0,0 +1,225 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class PosDetectorScanTest extends NexusTest {
+	
+	private IWritableDetector<PosDetectorModel> detector;
+	
+	@Before
+	public void before() throws Exception {
+		PosDetectorModel model = new PosDetectorModel(3);
+		model.setTimeout(100); // TODO remove
+		detector = (IWritableDetector<PosDetectorModel>) dservice.createRunnableDevice(model);
+		assertNotNull(detector);
+	}
+	
+	@After
+	public void after() throws Exception {
+		File parentDir = output.getParentFile();
+		String fileName = output.getName().substring(0, output.getName().indexOf('.'));
+		File outputDir = new File(parentDir, fileName);
+		for (File file : outputDir.listFiles()) {
+//			file.delete();
+		}
+//		outputDir.delete();
+	}
+	
+	@Test
+	public void testPosScan() throws Exception {
+		final int[] scanShape = new int[] { 8, 5 };
+		IRunnableDevice<ScanModel> scanner = createGridScan(detector, output, false, scanShape);
+		// add a the UniqueKeyChecker as a position listener to check the unique key is written at each point
+		((IPositionListenable) scanner).addPositionListener(
+				new UniqueKeyChecker(scanner.getModel().getFilePath(), scanShape));
+		
+		assertScanNotFinished(getNexusRoot(scanner).getEntry());
+		
+		scanner.run(null);
+		
+		checkNexusFile(scanner, scanShape);
+	}
+	
+	private static class UniqueKeyChecker implements IPositionListener {
+		
+		private final int[] scanShape;
+		private final String filePath;
+		private ILazyDataset uniqueKeysDataset;
+		
+		UniqueKeyChecker(String filePath, final int[] scanShape) {
+			this.filePath = filePath;
+			this.scanShape = scanShape;
+		}
+		
+		@Override
+		public void positionMovePerformed(PositionEvent event) throws ScanningException {
+			checkUniqueKeyWritten(event.getPosition());
+		}
+
+		@Override
+		public void positionPerformed(PositionEvent event) throws ScanningException {
+			checkUniqueKeyWritten(event.getPosition());
+		}
+		
+		private IDataset getUniqueKeysDataset() throws ScanningException {
+			try {
+				ILazyDataset uniqueKeysDataset = null;
+				if (uniqueKeysDataset == null) {
+					NexusFile nf = fileFactory.newNexusFile(filePath);
+					nf.openToRead();
+
+					TreeFile nexusTree = NexusUtils.loadNexusTree(nf);
+					nf.close();
+					NXroot root = (NXroot) nexusTree.getGroupNode();
+					NXentry entry = root.getEntry();
+
+					NXcollection solsticeScanCollection = entry.getCollection(GROUP_NAME_SOLSTICE_SCAN);
+					NXcollection keysCollection = (NXcollection) solsticeScanCollection.getGroupNode(GROUP_NAME_KEYS);
+					DataNode dataNode = keysCollection.getDataNode(FIELD_NAME_UNIQUE_KEYS);
+					uniqueKeysDataset = dataNode.getDataset();
+				}
+				
+//				((IDynamicDataset) uniqueKeysDataset).refreshShape();
+				return uniqueKeysDataset.getSlice(); // it's only a small dataset, so this is ok
+			} catch (Exception e) {
+				throw new RuntimeException(e);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)